src/librustc_mir/transform/unreachable_prop.rs

```diff
@@ -0,0 +1,37 @@
+//! A pass that propagates the unreachable terminator of a block to its predecessors
+//! when all of their successors are unreachable. This is achieved through a
+//! post-order traversal of the blocks.
+
+use rustc::ty::TyCtxt;
+use rustc::mir::*;
+use rustc_data_structures::fx::FxHashSet;
+use crate::transform::{MirPass, MirSource};
+
+pub struct UnreachablePropagation;
+
+impl MirPass<'_> for UnreachablePropagation {
+    fn run_pass<'tcx>(&self, _tcx: TyCtxt<'tcx>, _: MirSource<'tcx>, body: &mut Body<'tcx>) {
+        let mut unreachable_blocks = FxHashSet::default();
+
+        for (bb, bb_data) in traversal::postorder(body) {
+            if let Some(terminator) = &bb_data.terminator {
+                if let TerminatorKind::Unreachable = terminator.kind {
+                    unreachable_blocks.insert(bb.clone());
+                } else {
+                    let mut bb_successors = terminator.successors().peekable();
+
+                    if bb_successors.peek().is_some() {
```

Comment:
Actually, why this condition? Definitionally, a terminator which cannot end up anywhere means that it is unreachable.

Reply:
Do you mean the else branch and that it is redundant as we have the successors check?

Reply:
all() is true if it's empty alas, so we added a check that it wasn't empty first. Is there a cuter way to say all_but_not_empty()?

Reply:
I mean that the unwritten else-branch here could do the same as the if-branch, and so the `if bb_successors.peek().is_some()` is unnecessary.

Reply:
@gilescope But why do we want to differentiate between the empty and non-empty case? By definition, a branch is only reachable if it has any reachable successors.

Reply:
Never mind; this is wrong since this would transform e.g. `Abort` to `Unreachable` (there is a successor of sorts in this case, it's just not encoded in MIR...). Can you add a comment explaining why the check exists?